# Введение в машинное обучение

## Неделя-1. Знакомство с анализом данных и машинным обучением:
### Введение. Примеры задач. Логические методы: решающие деревья и решающие леса.

### Lab-1-1. Предобработка данных в Pandas

**Загрузите датасет `titanic.csv` и найдите ответы на вопросы:**

**1. Какое количество мужчин и женщин ехало на корабле?**

В качестве ответа приведите два числа через пробел.

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('data\\titanic.csv', index_col='PassengerId')

In [3]:
data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

In [4]:
data.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [5]:
data.groupby('Sex')['Name'].count()

Sex
female    314
male      577
Name: Name, dtype: int64

In [6]:
data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

`577` `314`

**2. Какой части пассажиров удалось выжить? Посчитайте долю выживших пассажиров.**

Ответ приведите в процентах (число в интервале от 0 до 100, знак процента не нужен), округлив до двух знаков.

In [7]:
survived = data.groupby('Survived')['Name'].count()
survived

Survived
0    549
1    342
Name: Name, dtype: int64

In [8]:
survived[1] / len(data)

0.3838383838383838

`38.38`

**3. Какую долю пассажиры первого класса составляли среди всех пассажиров?**

Ответ приведите в процентах (число в интервале от 0 до 100, знак процента не нужен), округлив до двух знаков.

In [9]:
p_class = data.groupby('Pclass')['Name'].count()
p_class

Pclass
1    216
2    184
3    491
Name: Name, dtype: int64

In [10]:
data['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [11]:
p_class[1] / len(data)

0.24242424242424243

`24.24`

**4. Какого возраста были пассажиры? Посчитайте среднее и медиану возраста пассажиров.**

В качестве ответа приведите два числа через пробел.

In [12]:
data['Age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

`29.7` `28`

**5. Коррелируют ли число братьев/сестер/супругов с числом родителей/детей?**

Посчитайте корреляцию Пирсона между признаками SibSp и Parch.

In [13]:
np.corrcoef(data['SibSp'], data['Parch'])

array([[1.       , 0.4148377],
       [0.4148377, 1.       ]])

In [14]:
data[['SibSp', 'Parch']].corr()

SibSp     Parch
SibSp  1.000000  0.414838
Parch  0.414838  1.000000

`%timeit np.corrcoef(data['SibSp'], data['Parch'])`

186 µs ± 13.1 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

`%timeit data[['SibSp', 'Parch']].corr()`

1.55 ms ± 74.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

0.41 < 0.5

слабая корреляция

`0.41`

**6. Какое самое популярное женское имя на корабле? Извлеките из полного имени пассажира (колонка Name) его личное имя (First Name).**

Это задание — типичный пример того, с чем сталкивается специалист по анализу данных. 

Данные очень разнородные и шумные, но из них требуется извлечь необходимую информацию. 

Попробуйте вручную разобрать несколько значений столбца Name и выработать правило для извлечения имен, а также разделения их на женские и мужские.

In [15]:
data['Last_name'] = data['Name'].apply(lambda x: x.split(', ')[0])
data['First_name'] = data['Name'].apply(lambda x: x[x.find('(') + 1:].split()[0] if x.find('(') != -1 else x[x.find('.') + 1:].split()[0])
data[data['Sex'] == 'female']['First_name'].value_counts().head()

Anna         14
Mary         13
Elizabeth    11
Margaret     10
Alice         6
Name: First_name, dtype: int64

`Anna`